subsys/bluetooth/host/adv.c

```diff
@@ -1553,6 +1553,26 @@ void bt_hci_le_adv_set_terminated(struct net_buf *buf)
 	adv = bt_adv_lookup_handle(evt->adv_handle);
 	conn_handle = sys_le16_to_cpu(evt->conn_handle);
 
+#if (CONFIG_BT_ID_MAX > 1) && (CONFIG_BT_EXT_ADV_MAX_ADV_SET > 1)
+	bt_dev.adv_conn_id = adv->id;
+	for (int i = 0; i < ARRAY_SIZE(bt_dev.cached_conn_complete); i++) {
+		if (bt_dev.cached_conn_complete[i].valid &&
+		    bt_dev.cached_conn_complete[i].evt.handle == evt->conn_handle) {
+			if (atomic_test_bit(adv->flags, BT_ADV_ENABLED)) {
+				/* Process the cached connection complete event
+				 * now that the corresponding advertising set is known.
+				 *
+				 * If the advertiser has been stopped before the connection
+				 * complete event has been raised to the application, we
+				 * discard the event.
+				 */
+				bt_hci_le_enh_conn_complete(&bt_dev.cached_conn_complete[i].evt);
+			}
+			bt_dev.cached_conn_complete[i].valid = false;
```

Comment:
It still seems that in case that the application disables the advertising set after the connection complete event is received, but before the adv terminated event, then we don't clean up the cached connection complete event, as there won't be a adv terminated event.